src/Microsoft.Azure.ServiceBus.KeyVault/KeyVaultPlugin.cs

```diff
@@ -0,0 +1,195 @@
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.ServiceBus.Core;
+    using System.Security.Cryptography;
+    using System.IO;
+    using System.Text;
+
+    /// <summary>
+    /// Provides Azure KeyVault functionality for Azure Service Bus.
+    /// </summary>
+    public class KeyVaultPlugin : ServiceBusPlugin
+    {
+        private readonly string secretName;
+        private readonly string keyVaultEndpoint;
+        private KeyVaultSecretManager secretManager;
+
+        /// <summary>
+        /// Gets the name that is used to identify this plugin.
+        /// </summary>
+        public override string Name => "KeyVaultPlugin";
+
+        /// <summary>
+        /// Creates a new instance of an <see cref="KeyVaultPlugin"/>.
+        /// </summary>
+        /// <param name="encryptionSecretName">The name of the secret used to encrypt / decrypt messages.</param>
+        /// <param name="options">The <see cref="KeyVaultPluginOptions"/> used to create a new instance.</param>
+        public KeyVaultPlugin(string encryptionSecretName, KeyVaultPluginOptions options)
+        {
+            if (string.IsNullOrEmpty(encryptionSecretName))
+            {
+                throw new ArgumentNullException(nameof(encryptionSecretName));
+            }
+
+            if (options != null)
+            {
+                if (string.IsNullOrEmpty(options.KeyVaultClientId))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(options.KeyVaultClientId));
+                }
+                if (string.IsNullOrEmpty(options.KeyVaultEndpoint))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(options.KeyVaultEndpoint));
+                }
+                if (string.IsNullOrEmpty(options.KeyVaultClientSecret))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(options.KeyVaultClientSecret));
+                }
+            }
+            else
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            this.secretName = encryptionSecretName;
+            this.keyVaultEndpoint = options.KeyVaultEndpoint;
+            this.secretManager = new KeyVaultSecretManager(options.KeyVaultEndpoint, options.KeyVaultClientId, options.KeyVaultClientSecret);
+        }
+
+        /// <summary>
+        /// The action performed before sending a message to Service Bus. This method will load the KeyVault key and encrypt messages.
+        /// </summary>
+        /// <param name="message">The <see cref="Message"/> to be encrypted.</param>
+        /// <returns>The encrypted <see cref="Message"/>.</returns>
+        public override async Task<Message> BeforeMessageSend(Message message)
+        {
+            try
+            {
+                if (message.UserProperties.ContainsKey(Constants.InitializationVectorPropertyName) || message.UserProperties.ContainsKey(Constants.KeyNamePropertyName))
```

Comment:
This check is to see if the key vault properties are on the message. If they aren't then we dont know which key / initialization vector to use